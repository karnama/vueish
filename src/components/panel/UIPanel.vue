<template>
    <section :id="id"
             class="ui-panel rounded shadow-md bg-default relative transition-all"
             :class="{ 'hover:shadow-lg': !open }">
        <UIFadeTransition>
            <UILinearLoader v-if="loading" class="absolute rounded-t" />
        </UIFadeTransition>

        <header v-if="$slots.header || $slots.actions"
                class="w-full flex items-center px-12 relative transition-all">
            <h1 class="select-none flex-grow py-8"
                :class="{
                    'cursor-pointer': collapsible,
                }"
                @click="toggle">
                <slot name="header" />
            </h1>

            <div v-if="open" class="my-2">
                <slot name="actions" />
            </div>
        </header>

        <UIExpandTransition :appear="appear">
            <div v-show="open">
                <main class="w-full px-12 py-6 text-gray-700">
                    <slot />
                </main>

                <footer v-if="$slots.footer" class="py-8 px-12 flex items-center">
                    <slot name="footer" />
                </footer>
            </div>
        </UIExpandTransition>
    </section>
</template>

<script lang="ts">
import UIFadeTransition from '../transitions/UIFadeTransition.vue';
import UIExpandTransition from '../transitions/UIExpandTransition.vue';
import UILinearLoader from '../loader-linear/UILinearLoader.vue';
import LocalCache from '../../helpers/cache/LocalCache';
import { defineComponent, ref, computed, watch } from 'vue';
import type { SetupContext } from 'vue';
import { SetupArg, SetupReturn } from '../../types';

let cache: LocalCache;

//todo - update icon
export default defineComponent({
    name: 'UIPanel',

    components: {
        UIFadeTransition,
        UILinearLoader,
        UIExpandTransition
    },

    props: {
        /**
         * Boolean flag to indicate the panel should start closed.
         */
        closed: {
            type: Boolean,
            default: false
        },

        /**
         * Boolean flag to indicate if the panel is collapsible.
         */
        noCollapse: {
            type: Boolean,
            default: false
        },

        /**
         * Id used for navigation to the portion of the page
         * and as a namespace for saved closed status.
         */
        id: {
            type: String,
            required: false
        },

        /**
         * Boolean flag for transition on appear.
         */
        appear: {
            type: Boolean,
            default: false
        },

        /**
         * Boolean flag to indicate if loader is visible.
         */
        loading: {
            type: Boolean,
            default: false
        }
    },

    setup(props: SetupArg, ctx: SetupContext): SetupReturn {
        const collapsible = computed(() => ctx.slots.header ? !props.noCollapse : false);
        const open = ref(!ctx.slots.header || !collapsible.value ? true : !props.closed);
        watch(() => props.closed, () => open.value = !props.closed);

        if (props.id) {
            cache = new LocalCache('panel_' + String(props.id));

            if (collapsible.value) {
                open.value = cache.get<boolean>('isOpen', open.value )!;
            }
        }

        /**
         * Toggle the panel's closed state.
         */
        const toggle = (): void => {
            if (!collapsible.value) {
                return;
            }

            open.value = !open.value;

            if (cache) {
                cache.set('isOpen', open.value);
            }
        };

        return {
            open,
            toggle,
            collapsible
        };
    }
});
</script>
