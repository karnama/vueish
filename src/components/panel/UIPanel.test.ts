import { mount } from '@vue/test-utils';
import UIPanel from './UIPanel.vue';
import { nextTick } from 'vue';

describe('UIPanel', () => {
    it('should render', () => {
        const wrapper = mount(UIPanel, {
            props: {}
        });

        expect(wrapper).toBeDefined();
        expect(wrapper.exists()).toBe(true);
    });

    it('should display the actions slot when action content when given', async () => {
        const wrapper = mount(UIPanel, {
            props: {},
            slots: {
                actions: 'action content'
            }
        });
        await nextTick();

        expect(wrapper.html()).toContain('action content');
    });

    it('should display the header slot content when slot content given', async () => {
        const wrapper = mount(UIPanel, {
            props: {},
            slots: {
                header: 'header content'
            }
        });
        await nextTick();

        expect(wrapper.html()).toContain('header content');
    });

    it('should display the footer slot when footer content is given', async () => {
        const wrapper = mount(UIPanel, {
            props: {},
            slots: {
                footer: 'footer content'
            }
        });
        await nextTick();

        expect(wrapper.html()).toContain('footer content');
    });

    it('should display the default slot content', () => {
        const wrapper = mount(UIPanel, {
            props: {},
            slots: {
                default: 'default content'
            }
        });

        expect(wrapper.html()).toContain('default content');
    });

    it('should close/open by a prop', async () => {
        const wrapper = mount(UIPanel, {
            slots: {
                default: '<div id="default"/>'
            }
        });

        expect(wrapper.find('#default').isVisible()).toBe(true);
        await wrapper.setProps({
            closed: true
        });
        expect(wrapper.find('#default').isVisible()).toBe(false);
    });

    it('should close and open by clicking on the header', async () => {
        const wrapper = mount(UIPanel, {
            slots: {
                header: '<div id="header" />',
                default: '<div id="default" />'
            }
        });
        await nextTick();

        expect(wrapper.find('#default').isVisible()).toBe(true);
        await wrapper.find('#header').trigger('click');
        expect(wrapper.find('#default').isVisible()).toBe(false);
        await wrapper.find('#header').trigger('click');
        expect(wrapper.find('#default').isVisible()).toBe(true);

    });

    it('should disable the collapsible behavior when prop given', async () => {
        const wrapper = mount(UIPanel, {
            slots: {
                header: '<div id="header" />',
                default: '<div id="default" />'
            },
            props: {
                noCollapse: true
            }
        });
        await nextTick();

        expect(wrapper.find('#default').isVisible()).toBe(true);
        await wrapper.find('#header').trigger('click');
        expect(wrapper.find('#default').isVisible()).toBe(true);

    });
});
