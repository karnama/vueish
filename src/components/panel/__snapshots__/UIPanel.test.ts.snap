// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIPanel should render correctly 1`] = `
<section
  aria-expanded="true"
  class="ui-panel rounded shadow-md bg-default relative transition-all"
>
  <transition-stub>
    
    
  </transition-stub>
  <header
    class="w-full flex items-center px-12 relative transition-all no-blur"
  >
    <h1
      class="select-none flex-grow py-8 cursor-pointer"
    >
      
      header text
      
    </h1>
    <div
      class="my-2"
    >
      
      
    </div>
  </header>
  <transition-stub>
    
    <div
      class="no-blur"
    >
      <main
        class="w-full px-12 py-6 text-gray-700"
      >
        
        default slot text
        
      </main>
      <!--v-if-->
    </div>
    
  </transition-stub>
  <transition-stub>
    
    
  </transition-stub>
</section>
`;
